velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,221 @@
+
+    auto pool = context->pool();
+    std::vector<BufferPtr> resultIndices(numInputArrays);
+    std::vector<BufferPtr> resultNulls(numInputArrays);
+    std::vector<vector_size_t*> rawResultIndices(numInputArrays);
+    std::vector<uint64_t*> rawResultNulls(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      resultIndices[i] = allocateIndices(resultElementsSize, pool);
+      resultNulls[i] = AlignedBuffer::allocate<bool>(
+          resultElementsSize, pool, bits::kNotNull);
+      rawResultIndices[i] = resultIndices[i]->asMutable<vector_size_t>();
+      rawResultNulls[i] = resultNulls[i]->asMutable<uint64_t>();
+    }
+
+    // Final Row Vector
+    const auto finalArraySize = rows.end();
+    BufferPtr finalArrayIndices = allocateIndices(finalArraySize, pool);
+    BufferPtr finalArrayNulls =
+        AlignedBuffer::allocate<bool>(finalArraySize, pool, bits::kNotNull);
+    BufferPtr finalArrayOffsets = allocateOffsets(finalArraySize, pool);
+    BufferPtr finalArraySizes = allocateSizes(finalArraySize, pool);
+
+    auto rawFinalArrayOffsets = finalArrayOffsets->asMutable<vector_size_t>();
+    auto rawFinalArraySizes = finalArraySizes->asMutable<vector_size_t>();
+
+    // Create right offsets/indexes for the individual and final result arrays.
+    int rowIndex = 0;
+    int finalResultIndex = 0;
+    rows.applyToSelected([&](auto row) {
+      // Get the max size for that row.
+      auto maxRowSize = maxArraySizes[row];
+      rawFinalArrayOffsets[finalResultIndex] = rowIndex;
+
+      for (int arr = 0; arr < numInputArrays; arr++) {
+        auto arrayOffset = rawOffsets[arr][indices[arr][row]];
+        auto vectorSize = rawSizes[arr][indices[arr][row]];
+        for (int i = 0; i < maxRowSize; i++) {
+          if (i < vectorSize) {
+            rawResultIndices[arr][rowIndex + i] = arrayOffset++;
+          } else {
+            bits::setNull(rawResultNulls[arr], rowIndex + i, true);
+          }
+        }
+      }
+      rowIndex += maxRowSize;
+      rawFinalArraySizes[finalResultIndex] = maxRowSize;
+      finalResultIndex++;
+    });
+
+    // Now create result dictionary vectors.
+    std::vector<VectorPtr> resultDictionaryVectors(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      resultDictionaryVectors[i] = BaseVector::wrapInDictionary(
+          resultNulls[i],
+          resultIndices[i],
+          rowIndex,
+          baseVectors[i]->elements());
+    }
+
+    auto rowType = outputType->childAt(0);
+    auto rowVector = std::make_shared<RowVector>(
+        pool,
+        rowType,
+        BufferPtr(nullptr),
+        resultElementsSize,
+        resultDictionaryVectors);
+
+    // Now convert these to an Array
+    auto arrayVector = std::make_shared<ArrayVector>(
+        pool,
+        outputType,
+        finalArrayNulls,
+        finalResultIndex,
+        finalArrayOffsets,
+        finalArraySizes,
```

Comment:
I'm seeing that 'final' is used to refer to the array and 'result' to elements of the array. Would it be clearer to use 'result' and 'nestedResult' instead?

Reply:
Yes, agree tracking between final and result is confusing. Also the input is called rawSizes, rawOffsets, indices which makes the names short but confusing to track what they are. It might be better to call them inputSizes, inputOffsets etc.